### ISSUE TITLE
server got an invalid status for this operation

### ISSUE THREAD

**Issue by Sebi94nbg** [planetteamspeak/ts3phpframework#104]
Hey,
I try to delete a virtual server, which is already offline. Shouldn't this do the trick already?

```
$uri = "serverquery://serveradmin:v3rYSeCr3tP4s5w0Rd@10.100.1.134:10011/?server_id=136&blocking=1&nickname=TS3PHPFramework&use_offline_as_virtual#no_query_clients";

$TS3PHPFramework = new TeamSpeak3();
$virtualserver= $TS3PHPFramework->factory("$uri");
$virtualserver->delete();
```

Your `TeamSpeak3_Exception` is always returning this message: `server got an invalid status for this operation`

**Comment by snegrini** [comment]
Right, you can't select a server if it is offline, then just connect as host without selecting the virtual server, then pass the server id to the method `serverDelete()`.
Code example (tested myself):

```
    $serverId = 136;
    $uri = "serverquery://serveradmin:v3rYSeCr3tP4s5w0Rd@10.100.1.134:10011/&nickname=TS3PHPFramework";
    $host = TeamSpeak3::factory($uri);
    $host->serverDelete($serverId);  
```

**Comment by Sebi94nbg** [comment]
Thanks, this worked fine.

I've tried to delete the server all the time using the method `delete()`: https://docs.planetteamspeak.com/ts3/php/framework/class_team_speak3___node___server.html#a13bdffdd926f26b825ea57066334ff01

This method does not allow any parameters. However: It worked with your solution - I've just updated my code. Thanks a lot! :)

**Comment by snegrini** [comment]
Please don't close this yet, I think that `delete()`  is unusable in Node Server. A server must be online to be selected, but if it is online it can't be deleted!
Maybe we need a clarification from @ronindesign 😄

**Comment by ronindesign** [comment]
I see this, let me check and see what's going on. There may very well be a logic bug here.

**Comment by ronindesign** [comment]
I can confirm this as a bug.

The two relevant pieces are:

`TeamSpeak3_Node_Server::delete()`
```php
  /**
   * Deletes the virtual server.
   *
   * @return void
   */
  public function delete()
  {
    $this->getParent()->serverDelete($this->getId());
  }
```

`TeamSpeak3_Node_Host::serverDelete()`
```php
  /**
   * Deletes the virtual server specified by ID.
   *
   * @param  integer $sid
   * @return void
   */
  public function serverDelete($sid)
  {
    $this->serverListReset();
    $this->execute("serverdelete", array("sid" => $sid));
    TeamSpeak3_Helper_Signal::getInstance()->emit("notifyServerdeleted", $this, $sid);
  }
```

I would say `TeamSpeak3_Node_Server::delete()` is just a shorter path to auto-fill the server ID when deleting.

So two possible fixes might be:

- Remove `TeamSpeak3_Node_Server::delete()` since it's confusing and doesn't work as expected.
- Rework `TeamSpeak3_Node_Host::serverDelete()` to call `serverDeslect()`, `serverStop()`, etc as needed.

If one of these solutions sounds better let me know your thoughts on it.

If we can find a way that works for the first solution, I think it would be the most intuitive.

**Comment by snegrini** [comment]
In my opinion it is better to remove `TeamSpeak3_Node_Server::delete()` in order to prevent unintended actions while the virtualserver is selected.

**Comment by ronindesign** [comment]
Sounds good @snegrini, I agree and think a "top down" approach for managing virtual servers makes the most sense.
I.e., connect to the host, then use `TeamSpeak3_Node_Host` object to administer its virtual servers.

@Sebi94nbg I'm a bit confused by your code example, but I think I understand what you'd like:

In `TeamSpeak3::factory()`, for the URI example you give, the following option will be triggered:
```php
      if($uri->hasQueryVar("server_id"))
      {
        $node = $node->serverGetById($uri->getQueryVar("server_id"));
      }
```
In this case, `$node` changes from an instance of `TeamSpeak3_Node_Host` to to an instance of `TeamSpeak3_Node_Server`, when `TeamSpeak3_Node_Host::serverGetById()` returns given a valid `server_id`.

The issue / confusion in your example is that `$node->serverDelete()` does not exist for virtual server objects. I.e. `TeamSpeak3_Node_Server::serverDelete()` is not defined, only `TeamSpeak3_Node_Host::serverDelete()` exists.

Also, as you can see from the included signatures above, a parameter is required: `TeamSpeak3_Node_Host::serverDelete($server_id)` which follows since no virtual server has been selected. The same is true for `TeamSpeak3_Node_Host::serverStop($server_id)`

However, the rest of your logic would work fine, it just needs to be executed from the parent object (i.e. the host not virtual server object).

So, would something like this be close to what you're thinking?

_Delete a child virtual server from a parent host_
```php
$uri = "serverquery://serveradmin:v3rYSeCr3tP4s5w0Rd@10.100.1.134:10011";

$TS3PHPFramework = new TeamSpeak3();
$host = $TS3PHPFramework->factory("$uri"); // TeamSpeak3_Node_Host
$host->serverDelete(136);
```

_Stop child virtual server first before deleting if running_
```php
$uri = "serverquery://serveradmin:v3rYSeCr3tP4s5w0Rd@10.100.1.134:10011";

$TS3PHPFramework = new TeamSpeak3();
$host = $TS3PHPFramework->factory("$uri"); // TeamSpeak3_Node_Host
$host->serverStop(136);
$host->serverDelete(136);
```

I apologize if this is response is a bit overdone.
I just wanted to clarify and make sure you _do_ like the idea of deleting the virtual server from the host rather than directly from the virtual server (since your code shows the latter, but you agreed verbally to the former).

**Comment by svenpaulsen** [comment]
So... the problem in the initial report is that the parameter value for `use_offline_as_virtual` is missing. the URI should be:

```php
$uri = "serverquery://serveradmin:v3rYSeCr3tP4s5w0Rd@10.100.1.134:10011/?server_id=136&nickname=TS3PHPFramework&use_offline_as_virtual=1#no_query_clients";
```

I agree that TeamSpeak3_Node_Server::delete() did not make sense in previous versions as the virtual server needs to be stopped before it can be deleted. I've just pushed 4e75bf0c0797137f780c5f19950822beaa5ba97b which should resolve this issue.

**Comment by Sebi94nbg** [comment]
@ronindesign Yeah, sorry - I just wrote a quick comment without any documentation on my hand, but you understood it very good. :)

Your examples are perfect!

@svenpaulsen I'm sure, that your pushed change will solve the issue - it looks fine.

I'll test it asap.

**Comment by svenpaulsen** [comment]
I actually commited total crap and made a stupid copy & paste error there... fixed with 66838e18751c595bd0eb77d5cf89f0e371dc955d.
